### DISCUSSION TITLE
Disabling `simulateContract` on writes

### DISCUSSION THREAD

**Discussion by hammeiam** [wevm/wagmi#3209] [Q&A]
I'm attempting to use a `Privy`, `Wagmi` & `Viem` app with a simple contract that I've deployed on a local mainnet-fork via `anvil`. For some reason, I'm unable to perform a write because the simulate contract fails. I would like to skip the simulation and just attempt the write directly. Is this possible?

**Comment by POKENA7** [comment]
According to this documentation, you can use `writeContract` to skip `simulateContract`.
https://beta.wagmi.sh/core/api/actions/writeContract

But `simulateContract` failed, I think `writeContract` will fail as well.
There may be a problem with contracts or elsewhere.

Also, it is not recommended not to use `simulateContract`.
https://viem.sh/docs/contract/writeContract.html

**Reply by kquirapas** [reply]
Recently, `simulateContract` has been failing on me with this require statement `require(_amount <= _tokenUsdt.balanceOf(msg.sender), "Insufficient USDT balance.");` printed all necessary values and it should evaluate to `true`. I used `cast` from foundry to call the function itself and it works. I removed `simulateContract` and it works.

Is simulateContract simulating the TX differently than it would with a write TX?

**Reply by POKENA7** [reply]
Hmmm, didn't know such a strange thing could happen. I have not encountered the same problem.

According to the documentation, it simulates almost the same conditions, however I don't know the details.
https://viem.sh/docs/contract/simulateContract.html

**Comment by sl00k** [comment]
Hi, I want to disable it as well because I am getting

The contract function "buyTokenUsingBNB" reverted with the following reason:
Buy: Incorrect price

but the function doesn't take a price as input.

**Comment by kquirapas** [comment]
I found the solution. I didn't use `simulateContract` from `viem`, I used `prepareWriteContract` from `wagmi/core`

It was all over the place because I thought I needed to use `viem` with `wagmi` functions. You don't, it's just a peer dependency.

It turns out I just needed to use `wagmi/core` instead of `viem` functions.

Instead of using `viem` `writeContract` I used `wagmi/core` `prepareWriteContract` to get a `request` then use that for `wagmi/core` `writeContract`

Check docs here: https://wagmi.sh/core/actions/prepareWriteContract

**Comment by liho00** [comment]
I also want to disable that troublesome simulatecontract, made me debug for days
I got signature expired error, unable to proceed, but remix.ethereum working fine, damn

This should be an issue, not discussion haha

**Reply by kquirapas** [reply]
Use `@wagmi/core` `prepareWriteContract` instead over `viem`'s `simulateContract`.

**Comment by brasseale** [comment]
To add more confusion here, `viem`'s `simulateContract` function, for me, works fine with amounts less than 1 ether passed to the `value` parameter.

As soon as I pass a full 1 or >1 on the `value` parameter, it fails with "Insufficient Balance" error.

When I call the writeContract function directly, that works fine.

**Reply by kquirapas** [reply]
Use @wagmi/core prepareWriteContract instead over viem's simulateContract.

**Comment by miohtama** [comment]
I have the same problem that transaction simulation fails. [Wagmi 2.x has removed prepareWriteContract](https://wagmi.sh/core/guides/migrate-from-v1-to-v2#removed-preparewritecontract). What's the current equivalent of sending transaction without trying to simulate it with Wagmi?

I am looling into Wagmi's `writeContract` and there is some internal parameter which allows to bypass simulate, but the documentation and the code comments do not give a hint how to do this:

```
  let request
  if (__mode === 'prepared') request = rest
  else {
    const { request: simulateRequest } = await simulateContract(
      config,
      rest as any,
    )
    request = simulateRequest
  }
```

**Reply by antoinefarley** [reply]
@miohtama 
You can pass the `__mode` parameter to `writeContract` like this:
```typescript
writeContract(wagmiConfig, {
    ...
     __mode: 'prepared',
})
```

**Reply by beeb** [reply]
How is this documented nowhere in the `writeContract` docs??
